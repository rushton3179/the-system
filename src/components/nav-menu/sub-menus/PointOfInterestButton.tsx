import React, { FC } from 'react';
import clsx from 'clsx';

import { PointOfInterest } from '../../../App';
import useClickAndEnterKeyDown from '../../../hooks/useClickAndEnterKeydown';
import styles from './PointOfInterestButton.module.scss';

interface Props {
  /** Whether the component is currently visible. */
  isVisible: boolean;
  /** Point of interest object. */
  pointOfInterest: PointOfInterest;
  /** Whether the point of interest is currently being followed on screen. */
  isBeingFollowed: boolean;
  /** Click handler for when the button is clicked. */
  onPoiClick(): void;
}

/**
 * A button for points of interest. When Clicked the screen will follow the poi until
 * the button is clicked again or another poi is clicked.
 */
const PointOfInterestButton: FC<Props> = ({ isVisible, pointOfInterest, isBeingFollowed, onPoiClick }) => {
  const title = isBeingFollowed ? 'Click again to stop following.' : undefined;

  const [onClick, onEnter] = useClickAndEnterKeyDown(onPoiClick);
  return (
    <div key={pointOfInterest.display}>
      <div
        className={clsx(styles.navItem, { [styles.navItemFollowed]: isBeingFollowed })}
        onClick={onClick}
        onKeyDown={onEnter}
        role="button"
        tabIndex={isVisible ? 0 : undefined}
        title={title}
        aria-label={pointOfInterest.display}
      >
        {pointOfInterest.display}
      </div>
    </div>
  );
};

export default PointOfInterestButton;
