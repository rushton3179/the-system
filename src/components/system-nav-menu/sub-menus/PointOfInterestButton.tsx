import React, { FC } from 'react';
import clsx from 'clsx';

import { PointOfInterest } from '../../../App';
import useClickAndEnterKeyDown from '../../../hooks/useClickAndEnterKeydown';
import styles from './PointsOfInterestButton.module.scss';

interface Props {
  /** Point of interest object. */
  pointOfInterest: PointOfInterest;
  /** Whether the point of interest is currently being followed on screen. */
  isBeingFollowed: boolean;
  /** Click handler for when the button is clicked. */
  onPoiClick(): void;
}

/**
 * A button for points of interest. When Clicked the screen will follow the poi until
 * the button is clicked again or another poi is clicked.
 */
const PointOfInterestButton: FC<Props> = ({ pointOfInterest, isBeingFollowed, onPoiClick }) => {
  const title = isBeingFollowed ? 'Click again to stop following.' : undefined;

  const [onClick, onEnter] = useClickAndEnterKeyDown(onPoiClick);
  return (
    <div key={pointOfInterest.display}>
      <div
        className={clsx(styles.navItem, { [styles.navItemFollowed]: isBeingFollowed })}
        onClick={onClick}
        onKeyPress={onEnter}
        role="button"
        tabIndex={0}
        title={title}
        aria-label={pointOfInterest.display}
      >
        {pointOfInterest.display}
      </div>
    </div>
  );
};

export default PointOfInterestButton;
