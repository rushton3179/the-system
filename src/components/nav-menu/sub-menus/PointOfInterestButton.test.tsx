import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import PointOfInterestButton from './PointOfInterestButton';

describe('PointOfInterestButton', () => {
  test('it renders', () => {
    const poi = { display: 'Some point', ref: React.createRef() };
    render(<PointOfInterestButton pointOfInterest={poi} isBeingFollowed={false} onPoiClick={jest.fn()} />);
    const element = screen.getByText(poi.display);

    expect(element).toBeInTheDocument();
  });

  test('it matches snapshot', () => {
    const poi = { display: 'Some point', ref: React.createRef() };
    const { container } = render(
      <PointOfInterestButton pointOfInterest={poi} isBeingFollowed={false} onPoiClick={jest.fn()} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it changes orbits visible when Show Orbits clicked', () => {
    const poi = { display: 'Some point', ref: React.createRef() };
    const onPoiClick = jest.fn();

    render(<PointOfInterestButton pointOfInterest={poi} isBeingFollowed={false} onPoiClick={onPoiClick} />);

    userEvent.click(screen.getByText(poi.display));

    expect(onPoiClick).toBeCalledTimes(1);
  });

  test('it changes orbits visible when Show Orbits enter pressed', () => {
    const poi = { display: 'Some point', ref: React.createRef() };
    const onPoiClick = jest.fn();

    render(<PointOfInterestButton pointOfInterest={poi} isBeingFollowed={false} onPoiClick={onPoiClick} />);

    fireEvent.keyDown(screen.getByText(poi.display), { key: 'Enter', keyCode: 13 });

    expect(onPoiClick).toBeCalledTimes(1);
  });
});
