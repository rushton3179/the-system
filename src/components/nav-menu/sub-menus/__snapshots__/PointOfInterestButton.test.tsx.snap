// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointOfInterestButton it matches snapshot 1`] = `
<div>
  <div
    aria-label="pointsOfInterest.undefined"
    class="navItem"
    role="button"
  >
    pointsOfInterest.undefined
  </div>
</div>
`;
